#### Importing Libraries

In [33]:
import pandas as pd
import numpy as np
import os



### Importing Data

In [34]:
path = r'C:\\Users\\maad0\\OneDrive\\CAREERFOUNDRY NOTES\\Python Fundamentals for Data Analyst\\08-2023 Instacart Basket Analysis'

In [35]:
# Importing Products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [36]:
# Importing Orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data','Prepared Data','orders_wrangled.csv'), index_col = False)

In [65]:
# Step 1 Perform consitency check on df_prods dataframe.

df_prods.describe()

product_id      aisle_id  department_id        prices
count  49677.000000  49677.000000   49677.000000  49677.000000
mean   24850.194235     67.763110      11.728687      9.993164
std    14340.588602     38.316396       5.850651    453.592708
min        1.000000      1.000000       1.000000      1.000000
25%    12433.000000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.000000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [38]:
# Step 2 Run df.describe function on df_ords dataframe. 
## Share in markdown cell whether anything about the data looks off or should be investigated further.

df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  orders_time_of_day  days_since_prior_order  
count        3.421083e+06        3.421083e+06            3.214874e+06  
mean         2.776219e+00        1.345202e+01            1.111484e+01  
std          2.046829e+00        4.226088e+00            9.206737e+00  
min          0.000000e+00        0.000000e+00            0.000000e+00  
25%          1.000000e+00        1.000000e+01            4.000000e+00  
50%          3.000000e+00        1.300000e+01            7.000000e+00  
75%          5.000000e+00        1.600000e+01            1.500000e+01  
max          6.000000e+00        2.300000e+01            3.000000e+01

### Nothing in the df_ords dataframe looks strange.  

#### The frequency check on orders_day_of_week shows a Min of 0 days; Max of 6 days- which is accurate since there are 7-days in a week.

#### The orders_time_of_day is within a 24 hour period, with a Min of 0; Max of 2.5 hours.

#### The days_since_prior_order does not contain negative numbers.

## Creating a test dataframe

In [39]:
# Create a dataframe

df_test = pd.DataFrame()

In [40]:
# Create a mixed type column

df_test['mix'] = ['a','b',1,True]


In [41]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [42]:
# Check for mixed types

for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1) 
    if len (df_test[weird]) > 0:
        print (col)

mix


In [43]:
# Converting column data type from numeric to string

df_test['mix'] = df_test['mix'].astype('str')

In [44]:
# Finding Missing Values

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [45]:
# Create subset of dataframe containing missing value 'product_name'

df_nan = df_prods[df_prods['product_name'].isnull() ==True]

In [46]:
# Visualizing missing data

df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [47]:
# Compare the number of rows in current dataframe

df_prods.shape

(49693, 5)

In [48]:
# Create new dataframe

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [49]:
df_prods_clean.shape

(49677, 5)

In [50]:
# Using dropna as an alternative to drop all missing values

df_prods.dropna(subset = ['product_name'], inplace = True)

In [51]:
df_prods.shape

(49677, 5)

In [52]:
# Finding Duplicates


df_dups = df_prods_clean[df_prods_clean.duplicated()]




In [53]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [54]:
# Create a new dataframe that does not include dups


df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [55]:
df_prods_clean_no_dups.shape

(49672, 5)

## 4.5 Data Consistency Checks

In [67]:
# Perform consistency check on df_prods dataframe.

df_prods.describe()


product_id      aisle_id  department_id        prices
count  49677.000000  49677.000000   49677.000000  49677.000000
mean   24850.194235     67.763110      11.728687      9.993164
std    14340.588602     38.316396       5.850651    453.592708
min        1.000000      1.000000       1.000000      1.000000
25%    12433.000000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.000000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### Question 2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpre the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [56]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  orders_time_of_day  days_since_prior_order  
count        3.421083e+06        3.421083e+06            3.214874e+06  
mean         2.776219e+00        1.345202e+01            1.111484e+01  
std          2.046829e+00        4.226088e+00            9.206737e+00  
min          0.000000e+00        0.000000e+00            0.000000e+00  
25%          1.000000e+00        1.000000e+01            4.000000e+00  
50%          3.000000e+00        1.300000e+01            7.000000e+00  
75%          5.000000e+00        1.600000e+01            1.500000e+01  
max          6.000000e+00        2.300000e+01            3.000000e+01

>#### The Max value 99999.00000 looks off.  Further investigation is needed to determine whether this was a data- entry error. 

#### Question 3. Check for mixed-type data in your df_ords dataframe.

In [57]:
# Check for mixed types

for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print(col)

#### Question 4. If you find mixed-type data, fix it.  The column in question should contain observations of a single data type.

>There were no mixed-type data found in the df_ords dataframe

#### Question 5. Run a check for missing values in your df_ords dataframe .  Report your findings and propose an explanation for any missing values.

In [58]:
# Checking for missing values in df_ords dataframe.

df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
orders_time_of_day             0
days_since_prior_order    206209
dtype: int64

In [59]:
# Create a new dataframe

df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [60]:
df_ords_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  orders_time_of_day  days_since_prior_order  
0                         2                   8                     NaN  
11                        2                  11                     NaN  
26                        1                  14                     NaN  
39                        6                  11                     NaN  
45                        3                  12                     NaN  
...                     ...                 ...                     ...  
3420930                   4                  12                     NaN  
3420934                   3                  18                     NaN  
3421002                   6                  19                     NaN  
3421019                   1                  15                     NaN  
3421069                   3                  11                     NaN  

[206209 rows x 8 columns]

>>There are 206,209 missing values in the days_since_prior_order column. One of the following could have occurred: 1) We have new customers that have yet to place an order, or 2) Existing customers have only placed one order, resulting in the missing values for the days_since_prior_order column.

#### Question 6. Address the missing values using an appropriate method.  Explain why you used your method of choice.

> I chose not to remove the missing values as some customers may have only placed one order, resulting in the days_since_prior_order to be blank.  I think the best option is to flag these values vs. removing them.

#### Question 7. Run a check for duplicate values in your df_ords data. Report your findings and provide an explanation for any duplicates that you find.

In [61]:
df_ords_dup = df_ords[df_ords.duplicated()]

In [62]:
df_ords_dup

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, orders_time_of_day, days_since_prior_order]
Index: []

> There are no duplicates found in the df_ords dataframe.

#### Question 8. Address the duplicates using an appropriate method.  Explain why you used your method of choice.

> There are no duplicates to address, as none were found.

#### Question 9. Exporting final, cleaned dataframes-  df_prods and df_ords as .csv files

In [63]:
# Export df_ords as orders_checked

df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data','orders_checked.csv'))


In [64]:
# Export df_prods as products_checked

df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data','products_checked.csv'))